examples/ui/toolbar.rs

```diff
@@ -0,0 +1,279 @@
+        .run();
+}
+
+const BG_COLOR: Color = Color::rgb(0.1, 0.1, 0.1);
+const NORMAL_BUTTON: Color = Color::rgb(0.15, 0.15, 0.15);
+const HOVERED_BUTTON: Color = Color::rgb(0.25, 0.25, 0.25);
+const PRESSED_BUTTON: Color = Color::rgb(0.4, 0.4, 0.8);
+const SELECTED_BUTTON: Color = Color::rgb(0.35, 0.35, 0.75);
+
+/// System: Manages the click interaction and styling of all the button elements
+///
+/// Help wanted: I had to remove `Changed<Interaction>` from the query so that
+/// the buttons that weren't clicked would update.
+///
+/// Is there a way of listening to changes in `ResMut<State<ToolbarState>>`?
+fn button_interaction_system(
+    mut interaction_query: Query<(&Interaction, &ButtonTags), (Changed<Interaction>, With<Button>)>,
+    mut ev_toolbar_button: EventWriter<ToolbarEvent>,
+) {
+    for (interaction, button_tag) in &mut interaction_query {
+        match *interaction {
+            Interaction::Clicked => {
+                let message = match button_tag {
+                    ButtonTags::Stop => None,
+                    ButtonTags::Play => Some("Playing...".to_string()),
+                    ButtonTags::Pause => Some("Pause...".to_string()),
+                };
+
+                ev_toolbar_button.send(ToolbarEvent {
+                    tag: button_tag.clone(),
+                    message,
+                    interaction: interaction.clone(),
+                });
+            }
+            Interaction::Hovered => {
+                ev_toolbar_button.send(ToolbarEvent {
+                    tag: button_tag.clone(),
+                    message: None,
+                    interaction: interaction.clone(),
+                });
+            }
+            Interaction::None => {
+                ev_toolbar_button.send(ToolbarEvent {
+                    tag: button_tag.clone(),
+                    message: None,
+                    interaction: interaction.clone(),
+                });
+            }
+        }
+    }
+}
+
+fn toolbar_state_system(
+    mut ev_toolbar_button: EventReader<ToolbarEvent>,
+    mut toolbar_state: ResMut<State<ToolbarState>>,
+) {
+    for ev in ev_toolbar_button.iter() {
+        if ev.interaction == Interaction::Clicked {
+            if ev.tag != toolbar_state.current().selected {
+                println!("Toolbar state: Setting {:?} as selected", ev.tag);
+                toolbar_state
+                    .set(ToolbarState {
+                        selected: ev.tag.clone(),
+                    })
+                    .unwrap();
+            }
+        }
+    }
+}
+
+fn toolbar_style_system(
+    mut ev_toolbar_button: EventReader<ToolbarEvent>,
+    mut style_query: Query<(&mut Style, &mut UiColor, &ButtonTags), With<Button>>,
+    toolbar_state: Res<State<ToolbarState>>,
+) {
+    for ev in ev_toolbar_button.iter() {
+        for (mut style, mut color, tag) in &mut style_query {
+            // Update button colours
+            let this_button = ev.tag.clone() == tag.clone();
```

Comment:
Nit: You probably don't need to clone:
```suggestion
            let this_button = &ev.tag == tag;
```

Reply:
IIRC I ran into issues comparing references so I resolved this one by dereferencing as needed.